python/vast/vast/vast.py

```diff
@@ -37,78 +102,45 @@ async def start():
-            result[name].append(table)
-        return result
-
     @staticmethod
-    async def export_rows(
-        expression: str, mode: ExportMode, limit: int = -1
-    ) -> AsyncIterable[Dict[str, Any]]:
-        """Get a row wise view of the data
-
-        This method does not provide type information because it is using json
-        as export format. It is a temporary workaround for asynchronicity issues
-        with PyArrow, but we plan to have all VAST Python exports properly typed
-        and backed by Arrow."""
+    async def export(
+        expression: str, mode: ExportMode = ExportMode.HISTORICAL, limit: int = 100
+    ) -> AsyncIterable[TableSlice]:
+        """Executes a VAST and receives results as Arrow Tables."""
         if limit == 0:
             return
-        args = {}
-        match mode:
-            case ExportMode.CONTINUOUS:
-                args["continuous"] = True
-            case ExportMode.UNIFIED:
-                args["unified"] = True
-            case ExportMode.HISTORICAL:
-                pass
-        if limit > 0:
-            args["max_events"] = limit
-        proc = await CLI().export(**args).json(expression).exec()
-        while True:
-            if proc.stdout.at_eof():
-                break
-            line = await proc.stdout.readline()
-            if line.strip() == b"":
-                continue
-            line_dict = json.loads(line)
-            assert isinstance(line_dict, Dict)
-            yield line_dict
-
-        _, stderr = await proc.communicate()
-        if proc.returncode != 0:
-            logger.error(stderr.decode())
+        cmd = CLI().export(**VAST._export_args(mode, limit))
+        if expression == "":
+            cmd = cmd.arrow()
+        else:
+            cmd = cmd.arrow(expression)
+        proc = cmd.sync_exec()
+
+        def log():
+            id = secrets.token_hex(6)
+            export_str = f"export({expression}, {mode}, {limit})"
+            logger.debug(f"[export-{id}] Start logging for {export_str}")
+            for line in iter(proc.stderr.readline, b""):
+                logger.debug(f"[export-{id}] {line.decode().strip()}")
+            logger.debug(f"[export-{id}] Stop logging")
+
+        # TODO: don't use default thread pool here as it won't scale
+        t = asyncio.create_task(asyncio.to_thread(log))
+
+        try:
+            # VAST concatenates IPC streams for different types so we need to
+            # spawn multiple stream readers
+            while True:
+                logger.debug("starting new record batch stream")
+                async for batch in AsyncRecordBatchStreamReader(proc.stdout):
+                    yield batch
+        except Exception as e:
+            if isinstance(e, pa.ArrowInvalid):
+                logger.debug("completed processing stream of record batches")
+            else:
+                proc.kill()
+                await asyncio.wait_for(t, 3)
```

Comment:
> Assuming the wait gets a timeout in seconds.. Why 3?
3 because I wanted it to be more than 2 but less than 4 😄 

Joke aside:
> Is there a reason to KILL instead of TERM here?
We are already in an unexpected error case, so the there is no point in being graceful in the exit. We just want to get out of here 
> Shouldn't log be guaranteed to exit, so we could just do a regular wait?
the wrapping with a timeout comes from a general distrust in the current state as we are in an unexpected error. We know we failed, there is no point in taking the risk of hanging on top of it. I think log is guaranteed to exit, but there is nothing worth than debugging something that is hanging so I prefer to be safe and exit. I don't know what time to choose here. 3 seems more than enough for the logger to wrap up while failing relatively fast if things are hanging. I should provide a small explanation of this logic.

Reply:
```suggestion
                # Use SIGKILL as we are already in an unexpected error state, so
                # there is no point in trying a graceful exit.
                proc.kill()
                # We set a timeout to make sure that we don't hang while tearing
                # down the logging task. The duration of 3s is arbitrary and is
                # expected to be more than enough for the logger task to
                # complete.
                await asyncio.wait_for(t, 3)
                raise e
```